### PR TITLE
CORE-4597: Add optional topic deletion

### PR DESCRIPTION
**Pull request by benmcmahon100**

### CODE REVIEW

File: tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        val output =
+            listOf(
+                "kafka-topics.sh --command-config /tmp/working_dir/config.properties --bootstrap-server $address --list > $outputPath",
+                // The first half of this tests that there is more than one line in the file (an "empty" file still contains a newline
+                "c=\$(wc -l < $outputPath); if [ \$c -gt 1 ]; then $operation; fi"
+            ).joinToString(System.lineSeparator())
+
+        if (outputLocation != null) {
+            logger.info("Wrote to path $outputLocation")
```
Comment (davidcurrie) [inline tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt]: nit: at the point you log this you haven't written anything!

File: charts/corda/templates/topic-job.yaml [hunk 2]
```diff
@@ -79,6 +79,42 @@ spec:
           volumeMounts:
             - mountPath: /tmp/working_dir
               name: working-volume
+        {{- if .Values.kafka.cleanUp }}
+        - name: delete-topic-script
+          image: {{ include "corda.bootstrapImage" . }}
```
Comment (davidcurrie) [inline charts/corda/templates/topic-job.yaml:83]: `imagePullPolicy` and `resources` should go here too

File: charts/corda/templates/topic-job.yaml [hunk 3]
```diff
@@ -79,6 +79,42 @@ spec:
...
+          image: bitnami/kafka:3.1.0-debian-10-r89
+          imagePullPolicy: {{ .Values.imagePullPolicy }}
+          {{- include "corda.bootstrapResources" . | nindent 10 }}
+          command:
+            - /bin/bash
+            - -c
+          args:
+            - |
+              set -e
+              export KAFKA_HEAP_OPTS="-Xmx1G"
```
Comment (davidcurrie) [inline charts/corda/templates/topic-job.yaml]: seems like a bad thing to hard-code this if we're letting them configure the resources. Maybe `-XX:MaxRAMPercentage` would be a better choice?

File: tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import kotlin.io.path.createDirectories
+import kotlin.io.path.exists
+
+@CommandLine.Command(name = "delete", description = ["Generates deletion script for cleaning up topics"])
+class Delete(
+    private val writerFactory: (String) -> FileWriter = { file -> FileWriter(File(file)) },
+) : Runnable {
+    @CommandLine.Option(
+        names = ["-f", "--prefix"],
+        description = ["Set topic prefix for created topics"]
```
Comment (davidcurrie) [inline tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt]:
```suggestion
        description = ["Prefix for topics to delete"]
```

File: tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    )
+    var topicPrefix: String? = null
+    @CommandLine.Option(
+        names = ["-a", "--address"],
+        description = ["Bootstrap server address for topic create"],
```
Comment (davidcurrie) [inline tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt]:
```suggestion
        description = ["Bootstrap server address for topic deletion"],
```

File: charts/corda/values.yaml [hunk 6]
```diff
@@ -67,6 +67,8 @@ kafka:
   replicas: 3
   # -- prefix to use for Kafka topic names (to support the use of a single Kafka cluster by multiple Corda clusters)
   topicPrefix: ""
+  # -- Specifies whether we should delete the previous topics with the given prefix before trying to create new ones. Uses provided prefix, and NOPs if set to true, but no topics are found
```
Comment (davidcurrie) [inline charts/corda/values.yaml]: think this should explicitly state what happens when a prefix is not specified i.e. that it's going to try and delete everything...

File: charts/corda/values.yaml [hunk 7]
```diff
@@ -67,6 +67,8 @@ kafka:
   replicas: 3
   # -- prefix to use for Kafka topic names (to support the use of a single Kafka cluster by multiple Corda clusters)
   topicPrefix: ""
+  # -- Specifies whether we should delete the previous topics with the given prefix before trying to create new ones. Uses provided prefix, and NOPs if set to true, but no topics are found
+  cleanUp: false
```
Comment (davidcurrie) [inline charts/corda/values.yaml]: `cleanup`?

File: tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    )
+    var outputLocation: String? = null
+
+    companion object {
+        private val logger: Logger = LoggerFactory.getLogger(this::class.java)
+    }
+
+    private fun createFilter(): List<String> {
+        return if (topicPrefix != null) {
+            listOf("grep -ie '^$topicPrefix'")
```
Comment (davidcurrie) [inline tools/plugins/topic-config/src/main/kotlin/net/corda/cli/plugins/topicconfig/Delete.kt]: given that topic names are case sensitive, I don't think we should perform a case insensitive search on them

**Review by davidcurrie — CHANGES_REQUESTED** [review]

File: charts/corda/templates/topic-job.yaml (hunk 2 above)
Comment (benmcmahon100) [reply charts/corda/templates/topic-job.yaml:83]: Strictly speaking the cli doesn't need the resources - but may as well make it consistent as it makes it easier to fit the pod

File: charts/corda/templates/topic-job.yaml (hunk 3 above)
Comment (benmcmahon100) [reply charts/corda/templates/topic-job.yaml]: Possibly, but I'm unsure if that kafka option would be happy receiving that as a value. I think 1G per topic creation is easier as well as then it means a blocksize of 8 will require ~8G of ram. Using a percent makes that a lot less clear

File: charts/corda/values.yaml [hunk 9]
```diff
@@ -67,8 +67,8 @@ kafka:
   replicas: 3
   # -- prefix to use for Kafka topic names (to support the use of a single Kafka cluster by multiple Corda clusters)
   topicPrefix: ""
-  # -- Specifies whether we should delete the previous topics with the given prefix before trying to create new ones. Uses provided prefix, and NOPs if set to true, but no topics are found
-  cleanUp: false
+  # -- Specifies whether we should delete the previous topics with the given prefix before trying to create new ones. Uses provided prefix, or assumes no prefix if none provided, and NOPs if set to true, but no topics are found
```
Comment (davidcurrie) [inline charts/corda/values.yaml]: Just to make it really explicit...

```suggestion
  # -- Specifies whether existing topics with the given prefix should be deleted before trying to create new ones (deletes all existing topics if no prefix is given)
```

File: charts/corda/templates/topic-job.yaml (hunk 3 above)
Comment (davidcurrie) [reply charts/corda/templates/topic-job.yaml]: but you're now assuming that they're going to make at least 8GB of RAM available to the container...

**Review by davidcurrie — APPROVED** [review]
🎉

**Pull request merged**